### PR TITLE
upgrade neo-cli and all the local plugins to the latest version

### PR DESCRIPTION
**Pull request by Liaojinghui**
This pr will automatically update both the `neo-cli` and all installed `plugins` to the most latest version.

### CODE REVIEW

**Comment by Liaojinghui** [comment]
image

**Comment by cloud8little** [comment]
why the plugin is still in version 3.0.1?

a32068d04ad430061032b1b88c25de2

b7cf20765e484d7fd6f62545f5db15a

**Comment by Liaojinghui** [comment]
image
Fixed now, it is because I forgot to copy the .dll file as it is outside of the `PluginName` folder

**Comment by cloud8little** [comment]
The plugin program file xx.dll will be replaced now. But before upgrade, I have oracleservice.dll in the Plugins file, while without OracleService folder(which contain config.json), and after upgrade, other dependencies such as grpc filestorage.dll is added, and OracleService folder new generated, but the folder is still empty without config.json.

c9550887ebb85d81495a4c9a42f461f

**Comment by cloud8little** [comment]
> `g.json` too.

I think if there is no config.json, then you can unzip it from the upgrade file, if config.json already exist, no need to override it.

**Comment by Liaojinghui** [comment]
> I think if there is no config.json, then you can unzip it from the upgrade file, if config.json already exist, no need to override it.

that is exactly the current logic.

**Comment by cloud8little** [comment]
Regardless of win/linux, why the upgrader program must need a runtime folder to work?

fcb37e7e632cbf1b9ea2290295df9bf

**Comment by Liaojinghui** [comment]
> win/linux, why the upgrader program must need a runtime folder to work?

It rely on Neo, I guess it is a dependency of Neo

**Review by superboyiii — DISMISSED** [review]

**Comment by superboyiii** [comment]
@shargon @erikzhang  Merge?

**Comment by erikzhang** [comment]
Tested?

**Comment by superboyiii** [comment]
> Tested?

Yes

File: update/Program.cs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            {
+                response = DownloadFromAPI(version, "neo-modules", pluginName);
+            }
+            using (response)
+            {
+                using Stream stream = response.GetResponseStream();
+                using ZipArchive zip = new(stream, ZipArchiveMode.Read);
+                try
+                {
+                    var temp = Path.Combine(Path.GetTempPath());
```
Comment (shargon) [inline update/Program.cs:131]: Combine with what?
```suggestion
                    var temp = Path.GetTempPath();
```

**Comment by Liaojinghui** [comment]
@shargon please review

**Comment by Liaojinghui** [comment]
image

**Comment by Liaojinghui** [comment]
@shargon can you please review?

File: update/Update.cs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            try
+            {
+                var targetFile = newPath.Replace(sourcePath, targetPath);
+                // Avoid config.json and other .json files
+                if (Path.GetExtension(newPath) != ".json" || !File.Exists(targetFile))
+                    File.Copy(newPath, targetFile, true);
+            }
+            catch
+            {
+                // ignored
```
Comment (shargon) [inline update/Update.cs:191]: I think that we need to log it

File: update/Update.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            }
```
Comment (shargon) [inline update/Update.cs:192]:
```suggestion
            catch (Exception ex)
            {
                ConsoleHelper.Error(ex.ToString());
            }
```

Comment (shargon) [reply update/Update.cs:192]: I think that we need to log it

**Review by shargon — APPROVED** [review]

**Comment by erikzhang** [comment]
I think the upgrade will be a little difficult. For example, automatic updating of configuration files (while keeping the original settings). Consider whether to resync blocks.

**Comment by Liaojinghui** [comment]
> I think the upgrade will be a little difficult. For example, automatic updating of configuration files (while keeping the original settings). Consider whether to resync blocks.

This just makes the process of `downloading new version from github then unzip it and copy it to target folder` into a single command.....as to config and resync, user can take care them manually (they are doing these manually now, nothing has changed).

**Comment by erikzhang** [comment]
Then I don't think this upgrader is useful. Because if an upgrader is used, the user will have a false sense of "it'll do everything for me", which often leads to problems.

**Pull request closed without merging**